src/GraphQL/Resolver.hs

```diff
@@ -105,11 +113,11 @@ ok :: GValue.Value -> Result GValue.Value
 ok = pure
 
 
--- TODO instead of SelectionSet we want something like
--- NormalizedSelectionSet which has query fragments etc. resolved.
+-- TODO: It's not crystal clear that "SelectionSet" should be the entry point.
```

Comment:
Maybe, yea. I went through a similar thought process as you when I wrote the following comment: 

```
-- TODO not super hot on individual values having to be instances of
-- HasGraph but not sure how else we can nest either types or
-- (Object _ _ fields). Maybe instead of field we need a "SubObject"?
```

But the current architecture is the only one I can think of that allows arbitrary nesting in a nice way, and nice composing is (to me) a good indicator that the design is OK, even if not super intuitive.

A counter-example would be using `Result` in `resolveField` - I think I used a hammer to push a square peg through a round hole there, and it doesn't quite work.